### PR TITLE
 Implement support for the IRCv3 sts specification.

### PR DESCRIPTION
**Pull request by SadieCat**
This has been tested to the best of my abilities but there may still be some edge cases as the connection logic is kinda gnarly.

### CODE REVIEW

**Comment by TingPing** [comment]
It may take me a while to better review this one.

**Comment by SadieCat** [comment]
That's understandable. It's a big change.

**Comment by SadieCat** [comment]
Rebased this to fix the few merge conflicts that existed from my other SSL change being merged.

**Comment by SadieCat** [comment]
@TingPing Have you had a chance to review this yet

File: src/common/inbound.c [hunk 1]
```diff
@@ -1811,6 +1815,57 @@ inbound_cap_ls (server *serv, char *nick, char *extensions_str,
...
+		if (!g_strcmp0 (extension, "sts"))
+		{
+			stscap = sts_parse_cap (value);
+			if (serv->use_ssl)
+			{
+				/* Retrieve the duration; if its valid then store/update it. */
+				stsvaluestr = (const char*) g_hash_table_lookup (stscap, "duration");
+				if (stsvaluestr)
+				{
+					stsvalue = atoi (stsvaluestr);
```
Comment (TingPing) [inline src/common/inbound.c:1829]: Could you add logging in these cases where parsing things fail. `g_debug ()`

File: src/common/inbound.c [hunk 2]
```diff
@@ -1811,6 +1815,57 @@ inbound_cap_ls (server *serv, char *nick, char *extensions_str,
...
+			{
+				stsvaluestr = (const char*) g_hash_table_lookup (stscap, "port");
+				if (stsvaluestr)
+				{
+					stsvalue = atoi (stsvaluestr);
+					if (stsvalue > 0 && stsvalue < 65536)
+					{
+						/* We are connecting on plain text and have found a valid STS profile. */
+						serv->stsprofile = sts_new ();
+						strcpy (serv->stsprofile->host, serv->hostname);
```
Comment (TingPing) [inline src/common/inbound.c:1843]: I don't want `strcpy` anywhere. `g_strlcpy()` usually is the better fit.

File: src/common/sts.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	profiles = g_slist_append (profiles, profile);
+}
+
+void sts_update_expiry (const char *host, time_t newexpiry, struct sts_profile *incomplete_profile)
+{
+	struct sts_profile *profile;
+
+	profile = incomplete_profile ? incomplete_profile : sts_find (host);
+	if (profile)
+		profile->expiry = time (NULL) + newexpiry;
```
Comment (TingPing) [inline src/common/sts.c:150]: This can be larger than `profile->expiry`, no validation is done.

File: src/common/sts.c [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+{
+	char buf[256];
+	int fh;
+	struct sts_profile *profile;
+
+	fh = hexchat_open_file (STS_CONFIG_FILE, O_RDONLY, 0, 0);
+	if (fh < 0)
+		return; /* Filesystem not readable. */
+
+	while (waitline (fh, buf, sizeof buf, FALSE) != -1)
```
Comment (TingPing) [inline src/common/sts.c:64]: Same comment that I want GIO based IO for reading files.

You can use [`g_data_input_stream_read_line()`](https://docs.gtk.org/gio/method.DataInputStream.read_line.html)

File: src/common/sts.c [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	{
+		profile = sts_new ();
+		if (sscanf (buf, "%s %u %ld", profile->host, &profile->port, &profile->expiry) != 3)
+		{
+			/* Malformed profile; drop it. */
```
Comment (TingPing) [inline src/common/sts.c:69]: `g_debug()`

File: src/common/outbound.c [hunk 6]
```diff
@@ -3447,6 +3447,10 @@ cmd_server (struct session *sess, char *tbuf, char *word[], char *word_eol[])
 		use_ssl_noverify = TRUE;
 		offset++;	/* args move up by 1 word */
 	}
+
+	/** If the server was previously used by a failed STS connection then free the incomplete profile. */
+	if (serv->stsprofile)
```
Comment (TingPing) [inline src/common/outbound.c:3452]: `g_clear_pointer()` is NULL safe

File: src/common/sts.c [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+	table = g_hash_table_new (g_str_hash, g_str_equal);
+	entries = g_strsplit (cap, ",", 0);
+	for (currentry = entries; *currentry; ++currentry)
+	{
+		value = strchr (*currentry, '=');
+		if (value)
+			g_hash_table_insert (table, g_strndup (*currentry, value - *currentry), g_strdup (value + 1));
+	}
+
+	g_free (entries);
```
Comment (TingPing) [inline src/common/sts.c:106]: Leak: `g_strfreev()`

File: src/common/sts.c [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+GHashTable *
+sts_parse_cap (const char* cap)
+{
+	char **entries, **currentry;
+	char *value;
+	GHashTable *table;
+
+	table = g_hash_table_new (g_str_hash, g_str_equal);
```
Comment (TingPing) [inline src/common/sts.c:97]: Leak: `g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);`

**Review by TingPing — CHANGES_REQUESTED** [review]
This is an initial review and this will take another round or so.

As I mentioned on IRC, I really dislike adding more blocking IO to this codebase. However this is common throughout hexchat so perhaps I should not block on that fact here.

A lot of little unsafe C nits I noticed so I will want to be a bit more detailed reviewing this later.

File: src/common/sts.c (hunk 3 above)
Comment (SadieCat) [reply src/common/sts.c:148]: How would you validate this in C? In C++ you would check if `std::numeric_limits<time_t>::max()` minus `newexpiry` is greater than the return value of `time(NULL)` but there's no TIME_MAX equivalent in C. The glib checked add functions don't work for time_t either.

**Comment by SadieCat** [comment]
Hey, sorry this took so long. I've rebased against the latest master and fixed all of the review comments.

**Comment by SadieCat** [comment]
If someone else wants to take this up then please feel free. I'm no longer interested in it.

**Pull request closed without merging**